HIP/hip-435.md

```diff
@@ -0,0 +1,178 @@
+The goal is to be able to externalise verbose information about transaction executions without bloating the main TransactionRecord object with potentially unnecessary information and requiring mirror nodes to always download and process that information.
+
+## Rationale
+
+There are use-cases requiring a detailed and verbose output for a HAPI transaction execution. Users of Hedera should have the option to query a mirror node for this information. Meanwhile, bloating the TransactionRecords with information that would be unnecessary in most cases is not an ideal option. It is a much better design decision to create sidecar record files that can be downloaded and processed only by interested parties.
+
+**Design Goal #1**
+Mirror nodes should not be required to download and process verbose transaction information. The sidecar transaction records will be externalised in separate record files.
+
+**Design Goal #2**
+One transaction can have 0, 1 or more than 1 sidecar records related to it located in 0, 1 or more record files.
+
+**Design Goal #3**
+It should be possible and relatively easy to add new types of sidecar transaction records.
+
+## User stories
+
+As a Hedera user, I want the to have the ability to access verbose information about transactions.
+
+As an operator of a mirror node, I want the ability to download and process only the most important record-related information that is externalised as part of the main RecordFile.
+
+As an operator of a mirror node, I want the ability to download sidecar record files containing all the information about records execution that a Hedera node can provide.
+
+## Specification
+
+Introduce a sidecar transaction record protobuf message and a sidecar record file protobuf message to enable tracking of additional information about records execution.
+
+### Record Stream File Changes
+
+A new version (V6) of the record stream file format will be created using the following protobufs:
+
+```
+message RecordFileObject {
+  Transaction transaction = 1;
+  TransactionRecord record = 2;
+}
+```
+
+`Transaction` and `TransactionRecord` are already defined as protobuf messages and are used by the Record Stream algorithm. No changes will be required for these messages.
+
+`RecordFileObjects` are serialized in record stream files.
+
+```
+message RecordStreamFile {
+  // Running Hash of all RecordFileObjects before writing this file.
+  bytes startObjectRunningHash = 1; 
+
+  // List of all the record file objects from that period.
+  repeated RecordFileObject fileObjects = 2; 
+
+  // Running Hash of all RecordFileObjects before closing this file.
+  bytes endObjectRunningHash = 3; 
+
+  // The block number associated with this period.
+  uint64 blockNumber = 4; 
+
+  // List of the hashes of all the Sidecar record files created for the same period.
+  // Allows multiple sidecar files to be linked to this RecordStreamFile
+  repeated bytes SideCarFileHash = 5; 
```

Comment:
Instead of just a hash we should add the sidecar index and type information here so that clients can choose which files to download. This is preferable to having type information encoded in the filename. Potentially the index can be omitted and be inferred from the list position.

```suggestion
  repeated SidecarMetadata sidecars = 5; 
```

```proto
enum SidecarType {
  ContractStateChange = 1;
  ContractAction = 2;
  // Other future categories
}

message SidecarMetadata {
  bytes hash = 1;
  long id = 2;
  repeated SidecarType types = 3;
}
```